jardin/database/__init__.py

```diff
@@ -15,42 +17,64 @@ class UnsupportedDriver(Exception): pass
+
+    _active_connections = {}  # For each db_name, it caches only ONE actively used connection. Active connection
+                              # can only be changed by a call to self.shuffle_connections(). The idea is to provide
+                              # connection stickiness during the lifetime of a session/task/job, etc.
 
     SUPPORTED_SCHEMES = ('postgres', 'mysql', 'sqlite', 'snowflake', 'redshift')
 
     @classmethod
     def connection(self, db_name):
-        if db_name not in self._connections:
-            self._connections[db_name] = self.build_connection(db_name)
-        return self._connections[db_name]
+        if db_name not in self._active_connections:
+            connections = self._connections.get(db_name)
+            if connections is None:
+                connections = self.build_connections(db_name)
+                self._connections[db_name] = connections
+            self._active_connections[db_name] = connections[0] if len(connections) == 1 else random.choice(connections)
+        return self._active_connections[db_name]
 
     @classmethod
-    def build_connection(self, name):
-        db = self.urls(name)
-        if db.scheme not in self.SUPPORTED_SCHEMES:
-            raise UnsupportedDriver('%s is not a supported driver' % db.scheme)
-        elif db.scheme == 'postgres' or db.scheme == 'redshift':
-            import jardin.database.drivers.pg as driver
-        elif db.scheme == 'mysql':
-            import jardin.database.drivers.mysql as driver
-        elif db.scheme == 'sqlite':
-            import jardin.database.drivers.sqlite as driver
-        elif db.scheme == 'snowflake':
-            import jardin.database.drivers.sf as driver
-
-        return driver.DatabaseConnection(db, name)
-
+    def build_connections(self, name):
+        connections = []
+        configs = self.db_configs(name)
+        for db in configs:
+            if db.scheme not in self.SUPPORTED_SCHEMES:
+                raise UnsupportedDriver('%s is not a supported driver' % db.scheme)
+            elif db.scheme == 'postgres' or db.scheme == 'redshift':
+                import jardin.database.drivers.pg as driver
+            elif db.scheme == 'mysql':
+                import jardin.database.drivers.mysql as driver
+            elif db.scheme == 'sqlite':
+                import jardin.database.drivers.sqlite as driver
+            elif db.scheme == 'snowflake':
+                import jardin.database.drivers.sf as driver
+            connections.append(driver.DatabaseConnection(db, name))
+        return connections
 
     @classmethod
-    def urls(self, name):
-        if len(self._urls) == 0:
+    def db_configs(self, name):
+        if len(self._db_configs) == 0:
             config.init()
-        for (nme, url) in config.DATABASES.items():
-            if url:
-                self._urls[nme] = DatabaseConfig(url)
-        return self._urls[name]
+            for (nme, urls) in config.DATABASES.items():
```

Comment:
i think you mean to have this for loop outside the if

Reply:
Why outside? I think the original code had a bug about this. 

We only need to run the for-loop if the cache collection (old: self._urls; new: self._db_configs), is empty.

Reply:
ah got it, makes sense. reading this again.